app/controllers/application_controller.rb

```diff
@@ -2,9 +2,140 @@
 
 class ApplicationController < Sinatra::Base
 
-  configure do
-    set :public_folder, 'public'
-    set :views, 'app/views'
-  end
+	configure do
+		set :public_folder, 'public'
+		set :views, 'app/views'
+		enable :sessions
+		set :session_secret, "password_security"
+	end
+
+	get '/' do 
+		erb :index
+	end
+
+	get '/tweets' do 
+		if logged_in?
+			erb :'tweets/tweets'
+		else
+			redirect '/login'
+		end
+	end
+
+	get '/signup' do 
+		if self.logged_in?
+			redirect '/tweets'
+		else
+			erb :'users/create_user'
+		end
+	end
+
+	get '/login' do
+		if logged_in?
+			redirect '/tweets'
+		else
+			erb :'users/login'
+		end
+	end
+
+	get '/logout' do 
+		if logged_in?
+			session.clear
+			redirect '/login'
+		else
+			redirect '/'
+		end
+	end
+
+	get '/users/:slug' do 
+		@user = User.find_by_slug(params[:slug])
+		erb :'users/user_tweets'
+	end
+
+
+	post '/signup' do 
+		@user = User.new
+		@user.username = params[:username]
+		@user.email = params[:email]
+		@user.password = params[:password]
+		if @user.save
+			session[:id] = @user.id
+			redirect '/tweets'
+		else
+			redirect '/signup'
+		end
+	end
+
+	post '/login' do
+		@user = User.find_by(username: params[:username])
```

Comment:
Alright, even if you implement a ActiveRecord validation based on username by adding
validates :username, uniqueness: true
in users.rb, the tests still break in the same way when trying to validate by slug with my code above:
  1) ApplicationController Signup Page does not let a logged in user view the signup page
     Failure/Error: expect(last_response.location).to include('/tweets')
       expected nil to include "/tweets", but it does not respond to `include?`
     # ./spec/controllers/application_controller_spec.rb:71:in `block (3 levels) in <top (required)>'